modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/wal/ByteBufferWalIteratorTest.java

```diff
@@ -0,0 +1,263 @@
+import static org.apache.ignite.internal.pagemem.wal.record.WALRecord.RecordType.BTREE_META_PAGE_INIT_ROOT_V3;
+import static org.apache.ignite.internal.pagemem.wal.record.WALRecord.RecordType.BTREE_PAGE_INNER_REPLACE;
+import static org.apache.ignite.internal.pagemem.wal.record.WALRecord.RecordType.BTREE_PAGE_MERGE;
+import static org.apache.ignite.internal.pagemem.wal.record.WALRecord.RecordType.PARTITION_META_PAGE_DELTA_RECORD_V4;
+
+/**
+ *
+ */
+public class ByteBufferWalIteratorTest extends GridCommonAbstractTest {
+
+    /** Cache name. */
+    private static final String CACHE_NAME = "cache";
+
+    /** */
+    private IgniteEx ig;
+
+    /** */
+    private GridCacheSharedContext<Object, Object> sharedCtx;
+
+    /** */
+    private GridCacheContext<Object, Object> cctx;
+
+    /** */
+    private IgniteWriteAheadLogManager wal;
+
+    /** */
+    private RecordSerializer serializer;
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        super.beforeTest();
+
+        ig = startGrid(0);
+
+        ig.cluster().state(ClusterState.ACTIVE);
+
+        sharedCtx = ig.context().cache().context();
+
+        cctx = sharedCtx.cache().cache(CACHE_NAME).context();
+
+        wal = sharedCtx.wal();
+
+        RecordSerializerFactory serializerFactory = new RecordSerializerFactoryImpl(sharedCtx);
+
+        serializer = serializerFactory.createSerializer(RecordSerializerFactory.LATEST_SERIALIZER_VERSION);
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        stopAllGrids();
+
+        super.afterTest();
+    }
+
+    /** */
+    private void writeRecord(IgniteWriteAheadLogManager wal, RecordSerializer serializer, ByteBuffer byteBuffer,
+        WALRecord walRecord) throws IgniteCheckedException {
+
+        // make sure walpointer is set.
+        wal.log(walRecord);
+
+        serializer.writeRecord(walRecord, byteBuffer);
+    }
+
+    /** */
+    private static boolean dataEntriesEqual(DataEntry x, DataEntry y) {
+        if (x == y)
+            return true;
+
+        if (x == null || y == null)
+            return false;
+
+        return x.cacheId() == y.cacheId()
+            && x.op() == y.op()
+            && Objects.equals(x.key(), y.key());
+    }
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String gridName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(gridName);
+
+        cfg.setCacheConfiguration(new CacheConfiguration<>(CACHE_NAME));
+
+        cfg.setDataStorageConfiguration(
+            new DataStorageConfiguration()
+                .setDefaultDataRegionConfiguration(
+                    new DataRegionConfiguration()
+                        .setPersistenceEnabled(true)
+                )
+        );
+
+        return cfg;
+    }
+
+    /** */
+    @Test
+    public void test() throws Exception {
+        List<WALRecord.RecordType> skipTypes = Arrays.asList(
```

Comment:
Can we replace the `skipTypes` with `.filter(Objects::nonNull)` in the stream below?

Reply:
Method `RecordUtils::buildWalRecord` creates non null object for registered types and for some of them it returns an instance of `UnsupportedWalRecord`, which doesn't work in `wal.log()`. So, instead of defining `skipTypes` we can check that the returned object is not null and not an instance of `UnsupportedWalRecord`. Will update the logic this way.